File: app/frontend/src/app/game/components/2_navigasi_menu/2_navigasi_bawah/5_geopolitik/2_organisasi_internasional/1_organisasi_PBB/logic/python/AI_finance/TreasuryScanner.ts
/**
 * TreasuryScanner TS Bridge: Matches the logic in treasury_scanner.py
 */
export interface TreasuryAnalysis {
    can_afford: boolean;
    is_sustainable: boolean;
    impact_ratio: number;
    remaining_treasury: number;
    reason?: string;
}

export class TreasuryScanner {
    private threshold: number;

    constructor(threshold = 0.2) {
        this.threshold = threshold;
    }

    public scanAffordability(currentTreasury: number, membershipFee: number): TreasuryAnalysis {
        if (currentTreasury <= 0) {
            return {
                can_afford: false,
                is_sustainable: false,
                impact_ratio: 1.0,
                remaining_treasury: currentTreasury,
                reason: "Treasury is empty or in deficit."
            };
        }

        const impactRatio = membershipFee / currentTreasury;
        const canAfford = currentTreasury >= membershipFee;
        const isSustainable = impactRatio <= this.threshold;

        return {
            can_afford: canAfford,
            is_sustainable: isSustainable,
            impact_ratio: Number(impactRatio.toFixed(4)),
            remaining_treasury: currentTreasury - membershipFee
        };
    }

    public getFinancialRecommendation(currentTreasury: number, membershipFee: number, priorityLevel: 'low' | 'normal' | 'high' | 'critical' = 'normal'): string {
        const analysis = this.scanAffordability(currentTreasury, membershipFee);
        
        if (!analysis.can_afford) return "REJECT: Insufficient funds.";

        const ratio = analysis.impact_ratio;

        if (priorityLevel === 'critical') {
            return ratio < 0.5 ? "APPROVE: Critical necessity despite high cost." : "REJECT: Even at critical priority, cost is suicidal.";
        }

        if (ratio < 0.05) return "APPROVE: Negligible impact on treasury.";
        if (ratio < 0.15) return "APPROVE: Reasonable investment.";
        if (ratio < 0.3) return "HOLD: High impact, requires secondary verification.";
        
        return "REJECT: Financial risk too high.";
    }
}

File: app/frontend/src/app/game/components/2_navigasi_menu/2_navigasi_bawah/5_geopolitik/2_organisasi_internasional/1_organisasi_PBB/logic/python/AI_finance/treasury_scanner.py
class TreasuryScanner:
    """
    Scans and analyzes the national treasury to determine if an AI country 
    can afford and sustain international organization memberships.
    """
    
    def __init__(self, treasury_threshold=0.2):
        # AI will not join if fee takes more than 20% of current treasury by default
        self.treasury_threshold = treasury_threshold

    def scan_affordability(self, current_treasury, membership_fee):
        """
        Analyzes the immediate impact on the treasury.
        """
        if current_treasury <= 0:
            return {
                "can_afford": False,
                "reason": "Treasury is empty or in deficit.",
                "impact_ratio": 1.0
            }
            
        impact_ratio = membership_fee / current_treasury
        
        can_afford = current_treasury >= membership_fee
        is_sustainable = impact_ratio <= self.treasury_threshold
        
        return {
            "can_afford": can_afford,
            "is_sustainable": is_sustainable,
            "impact_ratio": round(impact_ratio, 4),
            "remaining_treasury": current_treasury - membership_fee
        }

    def get_financial_recommendation(self, current_treasury, membership_fee, priority_level="normal"):
        """
        Provides a recommendation based on financial health and priority.
        Priority levels: 'low', 'normal', 'high', 'critical'
        """
        analysis = self.scan_affordability(current_treasury, membership_fee)
        
        if not analysis["can_afford"]:
            return "REJECT: Insufficient funds."
            
        ratio = analysis["impact_ratio"]
        
        # Recommendation logic based on priority
        if priority_level == "critical":
            # If critical, AI will join even if it takes 50% of treasury
            return "APPROVE: Critical necessity despite high cost." if ratio < 0.5 else "REJECT: Even at critical priority, cost is suicidal."
            
        if ratio < 0.05:
            return "APPROVE: Negligible impact on treasury."
        elif ratio < 0.15:
            return "APPROVE: Reasonable investment."
        elif ratio < 0.3:
            return "HOLD: High impact, requires secondary verification."
        else:
            return "REJECT: Financial risk too high."
